Use base class name for reification of associations

andre1810 opened the pull request:
Use base class name for reification instead of class name to reify polymorphic models properly.

Refers to #594 but won´t fix all issues. If you have more `has_one` associations to an polymorphic model you won´t get the correct associated models because sub type reference is missing in query.

jaredbeck commented on lib/paper_trail/reifiers/belongs_to.rb:
```diff
@@ -39,7 +39,7 @@ def load_record(assoc, id, options, version)
         def load_version(assoc, id, transaction_id, version_at)
           assoc.klass.paper_trail.version_class.
-            where("item_type = ?", assoc.klass.name).
+            where("item_type = ?", assoc.klass.base_class.name).
```
Nice. I was hoping AR provided something like this `base_class`, but I couldn't find it.

jaredbeck commented:
This looks good André. Please add:

1. at least one test, perhaps based on https://gist.github.com/andre1810/3cc1cbfaa8344a8ad8a840fbb9f91890, but let's use existing test models if we can. How about `Person`, `Animal`, `Cat`, and `Dog`? Person has many pets?
2. a changelog entry, under 9.0.0 -> Fixed (once this is in `master` we can discuss a backport if needed)

> .. [this] won´t fix all issues. If you have more has_one associations to an polymorphic model you won´t get the correct associated models because sub type reference is missing in query.

Do we have a bug report demonstrating that? Is there a gist I can run? Should we open a new issue for that or continue discussing it in https://github.com/airblade/paper_trail/issues/594?

andre1810 commented:
@jaredbeck I added a test to verify the changes regarding `has_many` associations and another one to show the still open issue #594 with `has_one` associations which is currently skipped. If you don´t want to include the issue I can remove it but I think it will help solving the issue later. Maybe we should keep #594 open because it´s not fixed at all.

I also added a changelog entry. Backward compatibility would be nice as I currently use 8.1.1.

jaredbeck commented on spec/models/person_spec.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  describe "#cars and bicycles" do
+    it "can be reified", skip: "Known Issue #594" do
```
This is great, thanks for including it.

jaredbeck commented on spec/models/pet_spec.rb:
```diff
@@ -0,0 +1,45 @@
+    expect(last_version.animals.first.name).to(eq("Beethoven"))
+    expect(last_version.dogs.first.name).to(eq("Beethoven"))
+    expect(last_version.animals.second.name).to(eq("Sylvester"))
+    expect(last_version.cats.first.name).to(eq("Sylvester"))
```
This test reads well. :+1:

jaredbeck commented on spec/dummy_app/db/migrate/20110208155312_set_up_test_tables.rb:
```diff
@@ -36,17 +36,18 @@ def up
-      t.datetime   :another_timestamp
-      t.timestamps null: true
+      t.datetime   :another_timestamp, limit: 6
+      t.timestamps null: true, limit: 6
```
Ugh. MySQL's lack of Fractional Seconds Precision (FSP) has been a big headache on this project. I'm OK with this change, but can you remind me what datatype this produces in MySQL? Is it a `DATETIME(6)`? Does that mean six decimal points of FSP?

I wonder if, after this change, we can use less `timecop` in our tests .. but that's a story for another time. :smile:

jaredbeck commented:
Can you organize this into two commits, the first adding fractional seconds precision and the second doing everything else? If it's a pain, don't worry about it, I'll just squash it into one commit when I merge.
